plugins/dns/plugin.go

```diff
@@ -37,20 +47,46 @@ func setupDNS4(args ...string) (handler.Handler4, error) {
 		if DNSServer.To4() == nil {
 			return Handler4, errors.New("expected an DNS server address, got: " + arg)
 		}
-		dnsServers = append(dnsServers, DNSServer)
+		dnsServers4 = append(dnsServers4, DNSServer)
 	}
-	log.Infof("loaded %d DNS servers.", len(dnsServers))
+	log.Infof("loaded %d DNS servers.", len(dnsServers4))
 	return Handler4, nil
 }
 
-// Handler6 not implemented only IPv4
+// Handler6 handles DHCPv6 packets for the dns plugin
 func Handler6(req, resp dhcpv6.DHCPv6) (dhcpv6.DHCPv6, bool) {
-	// TODO add DNS servers for v6 to the response
+	decap, err := req.GetInnerMessage()
+	if err != nil {
+		log.Errorf("Could not decapsulate relayed message, aborting: %v", err)
+		return nil, true
+	}
+
+	oro := decap.GetOneOption(dhcpv6.OptionORO)
+	if oro == nil {
+		// A client MUST include an Option Request option in a Solicit, Request,
+		// Renew, Rebind, or Information-request message to inform the server
+		// about options the client wants the server to send to the client
+		log.Warn("No OptionRequestOption in request, dropping")
+		return nil, true
+	}
+
+	for _, opt := range oro.(*dhcpv6.OptRequestedOption).RequestedOptions() {
+		if opt == dhcpv6.OptionDNSRecursiveNameServer {
+			resp.UpdateOption(&dhcpv6.OptDNSRecursiveNameServer{NameServers: dnsServers6})
+			break
+		}
+	}
 	return resp, false
 }
 
 //Handler4 handles DHCPv4 packets for the dns plugin
 func Handler4(req, resp *dhcpv4.DHCPv4) (*dhcpv4.DHCPv4, bool) {
-	resp.Options.Update(dhcpv4.OptDNS(dnsServers...))
+	if opts := req.ParameterRequestList(); opts != nil && !opts.Has(dhcpv4.OptionDomainNameServer) {
```

Comment:
NIT: this can be replaced with the `IsOptionRequested` method:
https://github.com/insomniacslk/dhcp/blob/master/dhcpv4/dhcpv4.go#L432

Reply:
Not exactly. There's a difference between : 

- There is not ParameterRequest option in the message
- There is a ParameterRequest option, which does not contain OptionDomainNameServer

IsOptionRequested returns the same (false) in both cases, while this only matches the second case on purpose (see the comments right below and in the else block)

Reply:
How about adding the option only when explicitly requested? Why is it being added when there is no `ParameterRequest` option present?

Reply:
I based this decision on https://tools.ietf.org/html/rfc2131#section-3.5, specifically: 

> Not all clients require initialization of all parameters listed in
> Appendix A.  Two techniques are used to reduce the number of
> parameters transmitted from the server to the client. [...]  Second, in its
> initial DHCPDISCOVER or DHCPREQUEST message, a client may provide the
> server with a list of specific parameters the client is interested in.

Which I interpret, especially the "reduce the number" and "may provide", as saying that if there is no option, the server should send all non-default parameters

I agree that it's not especially clear and there is no MUST or SHOULD to give a definitive answer. But compared to DHCPv6 (where OptionRequestOption is MUST in SOLICIT/REQUEST/etc) it seems the intent is different

https://tools.ietf.org/html/rfc2131#section-1.3 has similar language to say the same thing slightly differently: 

>   Not all of these parameters are required for a newly initialized
>   client.  A client and server may negotiate for the transmission of
>   only those parameters required by the client or specific to a
>   particular subnet.

Reply:
Maybe a solution here is to reverse the logic of IsOptionRequested in the library, to return true for everything if there is no ParameterRequest, instead of false ?

Reply:
> Maybe a solution here is to reverse the logic of IsOptionRequested in the library, to return true for everything if there is no ParameterRequest, instead of false ?

Ok, sounds good to me.